Complex path Uniswap bridge

benesjan opened the pull request:
# Description

Aztec Connect Bridge for swapping on Uniswap v3 along complex paths.

This is the gas report after enabling gas optimizations based on EIP-1087 (sending 1 wei of each input/output ERC20 token to the bridge):
<img width="648" alt="Screen Shot 2022-06-29 at 9 53 00" src="https://user-images.githubusercontent.com/13470840/176383421-1cb9926d-b984-40fb-91ea-1828a2dc5ef2.png">

To reproduce this gas report run the following command:
`forge test --fork-url https://mainnet.infura.io/v3/9928b52099854248b3a096be07a6b23c --fork-block-number 14972000 --match-contract UniswapBridge --gas-report`


# Checklist:

- [x] I have reviewed my diff in github, line by line.
- [x] Every change is related to the PR description.
- [x] There are no unexpected formatting changes, or superfluous debug logs.
- [x] The branch has been rebased against the head of its merge target.
- [x] I'm happy for the PR to be merged at the reviewers next convenience.

LHerskind commented on src/bridges/uniswap/UniswapBridge.sol:
```diff
@@ -0,0 +1,368 @@
+     */
+    function _decodeMinPrice(uint64 _encodedMinPrice) internal pure returns (uint256 minPrice) {
+        // 21 bits significand, 5 bits exponent
+        uint64 significand = _encodedMinPrice >> 5;
```
Why are you doing this with `uint64`? The `minPrice` will overflow if larger than `uint64` because the values are not being cast. Use `uint256` instead.

benesjan replied:
You are right. Will fix. Thanks 👍

benesjan replied:
It seems that solidity compiler is smart enough that it knows that the result is uint256 and for this reason it doesn't overflow. Pretty surprised tbh.

LHerskind reviewed (changes requested):
Need test going from low decimal count -> high decimal count and opposite direction as well. Mosts tests seems to go from 18->18

benesjan commented on src/bridges/uniswap/UniswapBridge.sol:
```diff
@@ -0,0 +1,368 @@
+ * @title Aztec Connect Bridge for swapping on Uniswap v3
+ * @author Jan Benes (@benesjan on Github and Telegram)
+ * @notice You can use this contract to swap tokens on Uniswap v3 along complex paths.
+ * @dev Encoding of a path allows for up to 2 split paths and up to 3 pools (2 middle tokens) in a each split path.
```
Uniswap docs use a term split route but I will describe it here anyway for clarity.

benesjan commented:
@LHerskind in the changes since your last review I changed the min price formula to not divide the value with anything. This is because I realized that manipulating the min price cause unnecessary precision loss. See [this function](https://github.com/AztecProtocol/aztec-connect-bridges/blob/3f7121eb08a5ec7cb4835b726ce2216c334fec99/src/bridges/uniswap/UniswapBridge.sol#L219) how the min price is encoded (please don't judge me because of the while loop, I know it could be optimized by having a binary search with a lot of ifs but I decided that it's not important in this context - used on frontend + tests). I implemented the fuzz tests and I was inspecting whether the encoding causes any issues (see the low to high and high to low decimals tests) and it turned out that this encoding is sufficient. Precision loss seems to be low for all the relevant scenarios.

LHerskind commented on src/test/bridges/uniswap/UniswapBridgeUnit.t.sol:
```diff
@@ -0,0 +1,325 @@
+import {UniswapBridge} from "../../../bridges/uniswap/UniswapBridge.sol";
+import {IQuoter} from "../../../interfaces/uniswapv3/IQuoter.sol";
+
+contract UniswapBridgeUnitTest is Test {
```
Would be useful to add USDT as an asset used in the testing as it doesn't follow the IERC20 interface and at the same time have <18 decimals.

benesjan replied:
Addressed in d7023ac58dfcb9fed141b3d440b0146dea11e49a

LHerskind commented on src/test/bridges/uniswap/UniswapBridgeUnit.t.sol:
```diff
@@ -0,0 +1,325 @@
+        uint256 tokenInDecimals = 6;
+
+        uint256 quote = QUOTER.quoteExactInput(abi.encodePacked(USDC, uint24(500), WETH), swapAmount);
+        uint256 desiredMinPrice = quote / swapAmount;
```
Why do you compute the `desiredMinPrice` as this? Say we swap 1000e6 USDC to 1e18 eth. Then we would a minPrice of 1e18/1e9 = 1e9. To me, it would seem more intuitive if the price is defined as the amount of asset B you receive for every whole asset A you supply. This is in practice also what you are doing when you are multiplying before doing the encoding.

benesjan replied:
Desired minPrice is computed here this way so that the min amount out in convert(...) simply matches quote from the quoter. With this I test that the test doesn't revert with InsufficientAmountOut(...).

LHerskind commented on src/bridges/uniswap/UniswapBridge.sol:
```diff
@@ -0,0 +1,504 @@
+
+    /**
+     * @notice A function which encodes min price to the format used in this bridge.
+     * @param _minPrice - Minimum acceptable price multiplied by 10**tokenInDecimals
```
Also added a comment in one of the test related to this.

benesjan replied:
The initial motivation for for passing in _minPrice multiplied by 10**tokenInDecimals was that in cases where decimals of token in are higher than decimals of tokenOut computing _minPrice would get rounded to 0 (amountIn/amountOut would get rounded to 0). Multiplying amountIn by 10**tokenInDecimals before dividing it with amountOut solved this issue.

I acknowledge that it's a bad design decision since it's definitely something which should stay in the internal logic and not be exposed to contract users.

I addressed this issue in e7b217c967569085d6bedbf36ed47a3fb7d2d143 by completely changing the interface of the function. I think now it's pretty clear to use and it avoids the issue mentioned above.

benesjan commented on src/bridges/uniswap/UniswapBridge.sol:
```diff
@@ -0,0 +1,504 @@
+     * reason the following is not a security risk and makes convert(...) function more gas efficient.
+     * @dev Note: If any of the input or output tokens is ETH just pass in a 0 address.
+     */
+    function preApproveTokenPair(address _tokenIn, address _tokenOut) external {
```
Good point. Addressed in cd201f8

benesjan commented on src/bridges/uniswap/UniswapBridge.sol:
```diff
@@ -0,0 +1,504 @@
+     * @param _tokenOut - Output ERC20 token
+     * @return path - Decoded/deserialized path struct
+     */
+    function _decodePath(
```
Addressed in edc1b6b7ec937968add19a092864e9575b8ba946

benesjan commented on src/bridges/uniswap/UniswapBridge.sol:
```diff
@@ -0,0 +1,504 @@
+     * @dev Reverts when _minPrice is bigger than max encodeable value.
+     */
+    function encodeMinPrice(uint256 _minPrice) external pure returns (uint256 encodedMinPrice) {
+        // 2097151 = 2**21 --> this number and its multiples of 10 can be encoded without precision loss
```
Addressed in edc1b6b7ec937968add19a092864e9575b8ba946

benesjan commented on src/test/bridges/uniswap/UniswapBridgeE2E.t.sol:
```diff
@@ -0,0 +1,101 @@
+import {IQuoter} from "../../../interfaces/uniswapv3/IQuoter.sol";
+
+contract UniswapBridgeE2ETest is BridgeTestBase {
+    address public constant LUSD = 0x5f98805A4E8be255a32880FDeC7F6728C6568bA0;
```
Addressed in d7023ac58dfcb9fed141b3d440b0146dea11e49a
